src/lib.rs

```diff
@@ -0,0 +1,501 @@
+            reference_hash: None,
+        };
+        self.nft.internal_mint(token_id, env::predecessor_account_id(), Some(token_metadata));
+    }
+
+    #[payable]
+    pub fn deposit_stablecoin(&mut self, vault_id: String, amount: U128) -> Promise {
+        let mut vault = self.vaults.get(&vault_id).expect("Vault not found");
+        let nft_collection = self.get_nft_collection_for_vault(&vault_id);
+        assert!(self.owns_nft(&env::predecessor_account_id(), &nft_collection.to_string()), "Not authorized");
+
+        vault.stablecoin_balance += amount.0;
+        self.vaults.insert(&vault_id, &vault);
+
+        Promise::new(vault.stablecoin.clone()).function_call(
+            "ft_transfer_call".to_string(),
+            serde_json::to_vec(&serde_json::json!({
+                "receiver_id": env::current_account_id(),
+                "amount": amount,
+                "msg": format!("deposit:{}", vault_id)
+            })).unwrap(),
+            NearToken::from_yoctonear(1),
+            Gas::from_tgas(10),
+        )
+    }
+
+    pub fn borrow(&mut self, vault_id: String, collateral_amount: U128, borrow_amount: U128) -> Promise {
+        let mut vault = self.vaults.get(&vault_id).expect("Vault not found");
+        let nft_collection = self.get_nft_collection_for_vault(&vault_id);
+        assert!(self.owns_nft(&env::predecessor_account_id(), &nft_collection.to_string()), "Not authorized");
+
+        assert!(vault.stablecoin_balance >= borrow_amount.0, "Insufficient liquidity");
+
+        vault.collateral_balance += collateral_amount.0;
+        vault.stablecoin_balance -= borrow_amount.0;
+        vault.loans.push(env::predecessor_account_id());
+
+        self.vaults.insert(&vault_id, &vault);
+
+        Promise::new(vault.collateral_asset.clone()).function_call(
+            "ft_transfer_call".to_string(),
+            serde_json::to_vec(&serde_json::json!({
+                "receiver_id": env::current_account_id(),
+                "amount": collateral_amount,
+                "msg": format!("collateral:{}", vault_id)
+            })).unwrap(),
+            NearToken::from_yoctonear(1),
+            Gas::from_tgas(10),
+        ).then(
+            Promise::new(vault.stablecoin.clone()).function_call(
+                "ft_transfer".to_string(),
+                serde_json::to_vec(&serde_json::json!({
+                    "receiver_id": env::predecessor_account_id(),
+                    "amount": borrow_amount,
+                })).unwrap(),
+                NearToken::from_yoctonear(1),
+                Gas::from_tgas(10),
+            )
+        )
+    }
+
+    pub fn create_invite(&mut self, nft_collection: AccountId) -> String {
+        assert!(self.nft_collections.get(&nft_collection.to_string()).is_some(), "NFT collection not found");
+        let invite_code = hex::encode(env::sha256(&env::random_seed()));
+        self.invites.insert(&invite_code, &nft_collection.to_string());
+        invite_code
+    }
```

Comment:
Couldn't get this to work, but will worry about this later since it's lower priority. pushed to another branch with bugs here: https://github.com/Templar-Protocol/contract-mvp/pull/4